connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java

```diff
@@ -724,9 +696,19 @@ private static Object convertToConnect(Schema schema, JsonNode jsonValue) {
                     break;
             }
         }
-        typeConverter = TO_CONNECT_CONVERTERS.get(schemaType);
+
+        final JsonToConnectTypeConverter typeConverter = TO_CONNECT_CONVERTERS.get(schemaType);
         if (typeConverter == null)
-            throw new DataException("Unknown schema type: " + schema.type());
+            throw new DataException("Unknown schema type: " + schemaType);
+
+        if (jsonValue.isNull()) {
+            assert schema != null; // otherwise we would have thrown above because of unknown schema type
+            if (schema.defaultValue() != null)
+                return schema.defaultValue(); // any logical type conversions should already have been applied
+            if (schema.isOptional())
+                return null;
+            throw new DataException("Invalid null value for required " + schemaType +  " field");
+        }
```

Comment:
Actually, I now understand why the logic is where it is, and why the logical conversion doesn't need to be done. However, I still the the above logic needs to only be performed if the `schema` is not null.